cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -149,8 +172,45 @@ class ParquetScanTaskIterator {
   static Status InferColumnProjection(const parquet::FileMetaData& metadata,
                                       const std::shared_ptr<ScanOptions>& options,
                                       std::vector<int>* out) {
-    // TODO(fsaintjacques): Compute intersection _and_ validity
-    *out = internal::Iota(metadata.num_columns());
+    // fall back to no push down projection
+    auto fall_back = [&] {
+      *out = internal::Iota(metadata.num_columns());
+      return Status::OK();
+    };
+
+    if (options->projector == nullptr) return fall_back();
+
+    // get column indices
+    out->clear();
+    auto filter_fields = FieldsInExpression(options->filter);
+
+    SchemaManifest manifest;
+    if (!SchemaManifest::Make(metadata.schema(), nullptr,
+                              parquet::default_arrow_reader_properties(), &manifest)
+             .ok())
+      return fall_back();
+
+    for (const auto& schema_field : manifest.schema_fields) {
+      auto field = schema_field.field;
+      auto parquet_column_index = schema_field.column_index;
+
+      // Ignore nested fields.
+      if (field->type()->num_children() != 0 || parquet_column_index == -1) {
```

Comment:
The goal of using SchemaManifest is that we can project nested fields.

```
auto is_referenced = [] (const string& field_name) { ... };
std::vector<int> parquet_column_indices;
for (const auto& schema_field : manifest.schema_fields) {
  if (is_referenced(schema_field.field()->name()) {
    // The following function recurse a SchemaField's children if !is_leaf() or else it push_back column_index.
    AddColumnIndices(schema_field, &parquet_column_indices);
  }
}
```

Reply:
We might also want to inverse the loop, e.g. for (auto field_name: referenced_field_names) ... since it seems some of our users have very wide columns. Also applicable to the statistics extraction method in my patch.